### PR TITLE
feat: add initial dotnet-support

### LINKED ISSUE #726: .NET Core-Support
**What would you like to be added**:
Hi Syft-Team,
are there any plans to support generating SBOMs for .NET Core projects?

**Why is this needed**:
.NET Core is a widely used ecosystem.

**Additional context**:

### PR DESCRIPTION
**Pull request by ckotzbauer**
Signed-off-by: Christian Kotzbauer <git@ckotzbauer.de>

This PR adds initial support for .NET (Core). It focuses on the parsing of the `*.deps.json` files in the `bin/` folders of a compiled project. This does not add support for parsing NuGet-related files which are usually only available at source-level.

I tested the changes against a compiled .NET project and against a ASP.NET Core image with the compiled project inside.

This changes are done alongside the PR adding dart-support. Please let me know, if anything is missing, wrong or should be changed.

Fixes #726
Ref #373

### CODE REVIEW

File: syft/pkg/dotnet_deps_metadata.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+type DotnetDepsMetadata struct {
+	Name    string `mapstructure:"name" json:"name"`
+	Version string `mapstructure:"version" json:"version"`
+}
+
+func (m DotnetDepsMetadata) PackageURL(_ *linux.Release) string {
+	var qualifiers packageurl.Qualifiers
+
+	return packageurl.NewPackageURL(
+		"dotnet", // TODO: packageurl.TypeDotnet,
```
Comment (ckotzbauer) [inline syft/pkg/dotnet_deps_metadata.go:17]: Missing TypeDotnet

File: syft/pkg/language.go [hunk 2]
```diff
@@ -66,6 +68,9 @@ func LanguageByName(name string) Language {
 		return Rust
 	case packageurl.TypePub, string(Dart):
 		return Dart
+	case string(Dotnet): // TODO:
+		//case packageurl.TypeDotnet, string(Dotnet):
```
Comment (ckotzbauer) [inline syft/pkg/language.go:72]: Missing TypeDotnet

File: syft/pkg/type.go [hunk 3]
```diff
@@ -65,6 +67,9 @@ func (t Type) PackageURLType() string {
 		return "cargo"
 	case DartPubPkg:
 		return packageurl.TypePub
+	case DotnetPkg:
+		// TODO: return packageurl.TypeDotnet
+		return "dotnet"
```
Comment (ckotzbauer) [inline syft/pkg/type.go:72]: Missing TypeDotnet

File: syft/pkg/type.go [hunk 4]
```diff
@@ -104,6 +109,9 @@ func TypeByName(name string) Type {
 		return RustPkg
 	case packageurl.TypePub:
 		return DartPubPkg
+	//case packageurl.TypeDotnet:
+	case "dotnet": // TODO:
+		return DotnetPkg
```
Comment (ckotzbauer) [inline syft/pkg/type.go:114]: Missing TypeDotnet

**Review by ckotzbauer — COMMENTED** [review]
I will open a PR against the `package-url/packageurl-go` repo when the namings are approved here.

File: syft/pkg/cataloger/dotnet/cataloger.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package dotnet
+
+import (
+	"github.com/anchore/syft/syft/pkg/cataloger/common"
+)
+
+// NewDotnetDepsCataloger returns a new Dotnet cataloger object base on deps json files.
+func NewDotnetDepsCataloger() *common.GenericCataloger {
+	globParsers := map[string]common.ParserFn{
+		"**/*deps.json": parseDotnetDeps,
```
Comment (wagoodman) [inline syft/pkg/cataloger/dotnet/cataloger.go:10]: Should there be an extra dot here?
```suggestion
		"**/*.deps.json": parseDotnetDeps,
```

File: syft/pkg/dotnet_deps_metadata.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package pkg
+
+import (
+	"github.com/anchore/packageurl-go"
+	"github.com/anchore/syft/syft/linux"
+)
+
+type DotnetDepsMetadata struct {
```
Comment (wagoodman) [inline syft/pkg/dotnet_deps_metadata.go:8]: This is probably a good opportunity to raise up the additional raw information for each of the packages in the `*.deps.json`, such as:
```
      "type": "package",
      "serviceable": true,
      "sha512": "sha512-kHBB+QmosVaG6DpngXQ8OlLVVNMzltNITfsRr68Z90qO7dSqJ2EHNd8dtBU1u3AQQLqqFHOY0lfmbpexeH6Pew==",
      "path": "awssdk.core/3.7.10.6",
      "hashPath": "awssdk.core.3.7.10.6.nupkg.sha512"
```

Straight from https://github.com/dotnet/cli/blob/v2.0.0/Documentation/specs/runtime-configuration-file.md

> The type property of a dependency object defines what kind of entity satisfied the dependency. Possible values include project and package (further comments on dependency types below).
...Open Question: type is also present in the libraries section. We don't really need it in both. It's in both now because the lock file does that and we want the formats to be similar. Should we remove it?

> # libraries Section (.deps.json)
This section contains a union of all the dependencies found in the various targets, and contains common metadata for them. Specifically, it contains the type, as well as a boolean indicating if the library can be serviced (serviceable, only for package-typed libraries) and a SHA-512 hash of the package file (sha512, only for package-typed libraries.

From a high level it seems like `path`, `sha512`, `hashPath` could be valuable to capture. I'm not certain of the value of `serviceable` or `type`, but wouldn't be against including these fields in `DotnetDepsMetadata`.

File: syft/pkg/language.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ const (
 	Go              Language = "go"
 	Rust            Language = "rust"
 	Dart            Language = "dart"
+	Dotnet          Language = "dotnet"
```
Comment (wagoodman) [inline syft/pkg/language.go:23]: Open question: do we prefer `.net` or `dotnet` for string representations of these values?

File: syft/pkg/dotnet_deps_metadata.go (hunk 6 above)
Comment (ckotzbauer) [reply syft/pkg/dotnet_deps_metadata.go:8]: This sounds good. I can include them. Not sure about `serviceable` and `type` too. Type would always be `package` (there's an if-statement) and `serviceable` will then always be true...

File: syft/pkg/cataloger/dotnet/cataloger.go (hunk 5 above)
Comment (ckotzbauer) [reply syft/pkg/cataloger/dotnet/cataloger.go:10]: Yes, that's missing

File: syft/pkg/language.go (hunk 7 above)
Comment (ckotzbauer) [reply syft/pkg/language.go:23]: hm, I think `.net` can be used for `.NET Framework` and `.NET Core`. `dotnet` always just means `.NET Core`, so I would prefer `dotnet`.

File: syft/pkg/dotnet_deps_metadata.go (hunk 6 above)
Comment (ckotzbauer) [reply syft/pkg/dotnet_deps_metadata.go:8]: Added `path`, `sha512` and `hashPath`

File: syft/pkg/language.go (hunk 7 above)
Comment (wagoodman) [reply syft/pkg/language.go:23]: sounds great -- thanks for the context 👍

**Comment by wagoodman** [comment]
re: packageurl-go updates, [we use a fork](https://github.com/anchore/packageurl-go) of [package-url/packageurl-go](https://github.com/package-url/packageurl-go) for this reason -- so we can get the changes needed for syft in asap without waiting for upstream. We submit PRs to upstream as well, though the merge time can take a while.

Did you want to also create a PR in our fork of packageurl-go for the `TypeDotnet` update? (if not we can leave it as is too and get the cataloger in)

File: syft/pkg/cataloger/cataloger.go [hunk 8]
```diff
@@ -44,6 +45,7 @@ func ImageCatalogers(cfg Config) []Cataloger {
 		java.NewJavaCataloger(cfg.Java()),
 		apkdb.NewApkdbCataloger(),
 		golang.NewGoModuleBinaryCataloger(),
+		dotnet.NewDotnetDepsCataloger(),
```
Comment (wagoodman) [inline syft/pkg/cataloger/cataloger.go:48]: Question: does the `*.deps.json` file positively indicate installed software? or is it mostly indicative of dependencies that you would find in a git repo?

This may seem like an odd question, but let me provide a little context. The image catalogers only raise up packages which have a high confidence of being installed. So for example, a ruby `*.gemspec` file alone may be an indication of an installed package or could be found in the root of a git repo and there is no software "installed". The way we distinct this is by ensuring that the gemspec is within a "specifications" directory, which is where gemspecs are found when they have been installed by the ruby toolchain (`**/specifications/**/*.gemspec`).

I'm trying to understand if the `.deps.json` falls into one or both categories. Does it indicate declared dependencies (e.g. in a source repo)? Does it indicate installed software? Or both? (and if both, can we distinct the two cases clearly enough, say as we do with gemspecs?)

Thanks in advance!

**Review by wagoodman — COMMENTED** [review]
_(not in scope for this PR, just capturing a comment for sometime in the future after this PR is merged)_ 

We could capture [`*.runtimeconfig.json`](https://github.com/dotnet/cli/blob/v2.0.0/Documentation/specs/runtime-configuration-file.md#appnameruntimeconfigjson) in the package metadata if it's available (would require significant cataloger changes).

**Comment by wagoodman** [comment]
I think the [linter issues](https://github.com/anchore/syft/runs/6197926925?check_suite_focus=true) can be fixed with `make lint-fix`

File: syft/pkg/cataloger/cataloger.go (hunk 8 above)
Comment (ckotzbauer) [reply syft/pkg/cataloger/cataloger.go:48]: The *.deps.json file is generated from the compiler, so you won't find that file in a git-repo as it would be gitignored. Does this answer the question?

**Comment by ckotzbauer** [comment]
> I think the [linter issues](https://github.com/anchore/syft/runs/6197926925?check_suite_focus=true) can be fixed with `make lint-fix`

fixed

Is the CLI-Test failure related to these changes?

> Did you want to also create a PR in our fork of packageurl-go for the TypeDotnet update?

Yes, I will do a PR against the fork.

File: syft/pkg/cataloger/cataloger.go (hunk 8 above)
Comment (wagoodman) [reply syft/pkg/cataloger/cataloger.go:48]: I think it does yeah --it's distinct from nuget (a package manager which would be installations) and distinct from a declared project manifest (which would be outside install cases) but still seems strongly associated with something that has been compiled and run.

Thanks!

**Comment by wagoodman** [comment]
@ckotzbauer I merged the pURL PR, so you should be good to go there.

re: CLI tests, the failure seems related to:
```
    packages_cmd_test.go:236: expected package count of 20, but found 32
...

    packages_cmd_test.go:240: STDERR:
         [0000]  INFO could not identify distro
        [0000]  INFO cataloging image
        
    packages_cmd_test.go:241: COMMAND: /home/runner/work/syft/syft/snapshot/linux-build_linux_amd64/syft packages -o json -s squashed docker-archive:test-fixtures/cache/stereoscope-fixture-image-pkg-coverage-d5baa8a0e48b30716c65fda2fecff73c8a85bdc023917b7ccae9ae0903d6d94c.tar
```

Since the `coverageImage` now includes `dotnet` packages, the total number of packages expected has grown:
- https://github.com/ckotzbauer/syft/blob/feature/dotnet/test/cli/packages_cmd_test.go#L99.

The easiest way to iterate locally on the CLI tests is:
- `make clean-snapshot snapshot` to create a build
- `make cli` to run tests against the build

I went ahead an fixed the test in question in 009dd0afadfc9ac9f46125b8148b7db4f8fa5e28, so no action needed 🎉

**Comment by wagoodman** [comment]
Open question: though I've merged https://github.com/anchore/packageurl-go/pull/8 I'm currently trying to make certain I understand the right usage of package URL for a .NET package relative to what a *.deps.json describes. As I understand it a .deps.json has nothing to do with nuget. However, in terms of pURL that might not matter. That is, the information contained within deps.json simply describe .NET packages, which canonically is nuget. 

In short: should we be using the [pURL type `nuget`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#nuget) instead of the proposed `dotnet` pURL type? 

In thinking/reading through this more I think the answer is "yes" we should use "nuget", but wanted to raise this up for comment.

**Comment by ckotzbauer** [comment]
Great, thanks for fixing the CLI test.

>  I merged the pURL PR, so you should be good to go there.

I will update this PR.

> should we be using the [pURL type nuget](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#nuget) instead of the proposed dotnet pURL type?

No, I don't think so. Nuget is the most popular package-manager for dotnet, but there are alternatives (e.g. https://fsprojects.github.io/Paket/index.html). I always used Nuget before, but I think the result in deps.json would be the same with another tool, as there's no reference to "nuget" in this file.

**Comment by wagoodman** [comment]
I think that all makes sense. I'll note that since the pURL type is non-standard that there is risk of needing to change this in the future, but I think for now we can operate under the assumption for now that a pURL that doesn't automatically couple .NET projects with `nuget` is the way to go in this instance.

**Comment by wagoodman** [comment]
Useful JSON schema diff for reviewers
```diff
# diff schema/json/schema-3.2.2.json  schema/json/schema-3.2.3.json
279a280,295
>     "DotnetDepsMetadata": {
>       "required": [
>         "name",
>         "version"
>       ],
>       "properties": {
>         "name": {
>           "type": "string"
>         },
>         "version": {
>           "type": "string"
>         }
>       },
>       "additionalProperties": true,
>       "type": "object"
>     },
520a537,542
>         },
>         "digest": {
>           "items": {
>             "$ref": "#/definitions/Digest"
>           },
>           "type": "array"
690a713,715
>               "$ref": "#/definitions/DotnetDepsMetadata"
>             },
>             {
```

**Review by wagoodman — APPROVED** [review]
cc: @anchore/tools 
Looks like there are minor tweaks in grype testing needed -- I can confirm there is no apparent unexpected matching behavior with the stock matcher. We'll still need to add a .NET-specific matcher and DB namespace finder to link up GHSA matches in the future.

Overall, great work @ckotzbauer ! Thanks for taking the time on this ❤️

**Comment by ckotzbauer** [comment]
Oh, just because you posted the schema diff: I extended the Metadata as part of the review, but I think I forgot to regenerate the schema.

**Comment by wagoodman** [comment]
Latest diff:
```diff
# diff schema/json/schema-3.2.2.json  schema/json/schema-3.2.3.json
279a280,307
>     "DotnetDepsMetadata": {
>       "required": [
>         "name",
>         "version",
>         "path",
>         "sha512",
>         "hashPath"
>       ],
>       "properties": {
>         "name": {
>           "type": "string"
>         },
>         "version": {
>           "type": "string"
>         },
>         "path": {
>           "type": "string"
>         },
>         "sha512": {
>           "type": "string"
>         },
>         "hashPath": {
>           "type": "string"
>         }
>       },
>       "additionalProperties": true,
>       "type": "object"
>     },
520a549,554
>         },
>         "digest": {
>           "items": {
>             "$ref": "#/definitions/Digest"
>           },
>           "type": "array"
690a725,727
>               "$ref": "#/definitions/DotnetDepsMetadata"
>             },
>             {
```
